src/test/java/io/confluent/examples/streams/utils/PairSerializer.java

```diff
@@ -0,0 +1,89 @@
+/*
+ * Copyright Confluent Inc.
+ *
+ * Licensed under the Apache License, Version 2.0 (the "License");
+ * you may not use this file except in compliance with the License.
+ * You may obtain a copy of the License at
+ *
+ *    http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+package io.confluent.examples.streams.utils;
+
+import org.apache.kafka.common.serialization.Serializer;
+
+import java.io.ByteArrayOutputStream;
+import java.io.DataOutputStream;
+import java.io.IOException;
+import java.util.Map;
+
+public class PairSerializer<X, Y> implements Serializer<Pair<X, Y>> {
+
+  private final Serializer<X> serializerX;
+  private final Serializer<Y> serializerY;
+
+  public PairSerializer(final Serializer<X> serializerX, final Serializer<Y> serializerY) {
+    this.serializerX = serializerX;
+    this.serializerY = serializerY;
+  }
+
+  @Override
+  public void configure(final Map<String, ?> configs, final boolean isKey) {
+  }
+
+  @Override
+  public byte[] serialize(final String topic, final Pair<X, Y> pair) {
+    if (pair == null) {
+      return null;
+    }
+
+    try (
+      final ByteArrayOutputStream baos = new ByteArrayOutputStream();
+      final DataOutputStream out = new DataOutputStream(baos)
+    ) {
+      final byte[] serializedX = serializerX.serialize(topic, pair.x);
```

Comment:
Should this only be called if `pair.x != null`? (Similar next line.)

The issue is, that you enforce that `deserialize` will not be called if `serializedX==null`, ie, you enforce a `null -> null` decoding -- hence, you should also enforce a `null -> null` encoding.

Reply:
No, it shouldn't be called.

The logic that determines the serialization is `Serializer<X>` -- we don't know in `PairSerializer` if the former turns a `null` input into a `null` output.

> you enforce a `null -> null` decoding
Where? You mean in `PairDeserializer`? If so, the decoding in `PairDeserializer` is determined by the magic byte.

Reply:
> No, it shouldn't be called.

Isn't this what I am saying? Or did you mean it _should_ be called, even if `pair.x == null` ?

There are four cases (given that we don't know what `serializerX.serialize(topic, pair.x);` does):

 - `pair.x == null` and `serializedX  == null`
 - `pair.x != null` and `serializedX  == null`
 - `pair.x == null` and `serializedX  != null`
 - `pair.x != null` and `serializedX  != null`

For the second case, `pair.x != null` and `serializedX  == null` you would set the magic byte to "not exist" fro x (either zero or two, depending on `serializedY`) and the deserializer for x would not be called, resulting in `pair.x = null` after deserialization. This breaks the logic.

Hence, I would enforce `pair.x == null` and `serializedX  == null` by not calling `serializerX.serialize(topic, pair.x);` at all and tell uses that need to return non-null for `pair.x != null` -- in fact, I would even check this and throw to enforce case 4.

Case 2 and case 3 get you into a bad position.

If course, if uses only use case 3 and case 4 in their logic, it would be correct and enforcing null to null mapping disallows this pattern. So if you want to support 3 and 4, your current code is fine, and you can claim that it might be a user bug if they mix 1 and 2...

Reply:
We don't care whether or not `pair.x` or `pair.y` are null.  We only care whether their serialized forms are null or not, and the magic byte is needed to communicate this information.